### ISSUE TITLE
If the subscription's payment method is declined, begin dunning cycle

### ISSUE THREAD

**Issue by chenkc805** [sourcegraph/cody#1600]
Dunning requirements:

- During dunning, we email the billing owner every day for days 1-4. For days 5-7, we email everyone else on the team to add pressure.
- If the card decline is a soft decline, retry the card every day for those 7 days. If the card is a hard decline, do not retry the card again, but keep the email flow going. 
- Dunning email copy: https://docs.google.com/document/d/1DD6Afb2z21G0v2vPRonuol0aikIjeRm1RjJBVeCLZy4/edit#heading=h.kjwa9r7xiqq8

Resources:
- Details being hashed out here: https://docs.google.com/spreadsheets/d/1QXnrHkgGLXdbfHEhI_VjtDTy-mCtnK1qri3IMQu7rNk/edit#gid=1512994667
```[tasklist]
### Tasks
- [ ] https://github.com/sourcegraph/cody/issues/3539
- [ ] https://github.com/sourcegraph/cody/issues/1601
- [ ] https://github.com/sourcegraph/cody/issues/1605
```

**Comment by vdavid** [comment]
[This summary](https://sourcegraph.slack.com/archives/C05PCA92ATT/p1700074915169899) is also helpful

**Comment by chrsmith** [comment]
We need to setup a subscription to test this. Currently blocked on the inability to change the credit card within the subscription management UI.

**Comment by vdavid** [comment]
Moving this post-Feb because we won't have dunning issues in the beginning.

**Comment by chenkc805** [comment]
@vdavid brought up a good point in today's meeting that in the following situation:

- User puts in credit card on Feb 8
- We try to charge the card on Feb 15
- Card fails for whatever reason (insufficient funds, card lost/stolen, etc.)

Since we don't have dunning in place by that time, automatically churn the user to Cody Free, and send them the [churn email](https://docs.google.com/document/d/1DD6Afb2z21G0v2vPRonuol0aikIjeRm1RjJBVeCLZy4/edit#heading=h.5geczfoke7qe)

**Comment by chrsmith** [comment]
So we need to fix the linked ticket around showing things in the UI, but as far as the Cody Pro product supporting dunning, I just confirmed this is all working as we expected.

- When there is a payment failure, the subscription goes into a `past_due` state. We send an email to the customer notifying them of the problem.

![Image](https://github.com/sourcegraph/cody/assets/4029847/b8ae0e48-86ea-4ffb-94b2-c3126b4ebcb0)

- We then will retry that failed payment 3x times. First, 3-days later. Then 5-days after that. And one final time 7-days after that.

Each time the same "Payment Failed" email will be sent, with the same "Update payment method" call to action.

- After the 3rd and final retry fails, the subscription transitions into the `canceled` state. We then send an email notifying them.

![Image](https://github.com/sourcegraph/cody/assets/4029847/97032678-1ebf-4410-80ee-bae7334ec4c3)

> During dunning, we email the billing owner every day for days 1-4. For days 5-7, we email everyone else on the team to add pressure.

We do not do this. And instead all emails will be sent to the Stripe customer. (i.e. one person, not all of the team admins, and not all team members later.)

I think that's a fine starting point. And when we ship the team tier, we can modify this behavior. (Though I still don't think sending the 2nd or 3rd notice to all team members makes sense, since there is nothing those folks can actually do to resolve the problem.)

I'll create a new ticket to figure out exactly what we want to do there (e.g. sending all billing-related emails to all team admins, or maybe introduce a way to designate "billing admins" or something, etc.)

**Comment by rrhyne** [comment]
I believe all design issues are addressed here:

https://github.com/sourcegraph/cody/issues/3539

**Comment by vdavid** [comment]
The design is ready, we can build this!

**Comment by vdavid** [comment]
@chrsmith will get this done by EOW

**Comment by chrsmith** [comment]
😬 well, that obviously didn't happen.
<img width="275" alt="image" src="https://github.com/sourcegraph/cody/assets/4029847/5c60ca8e-b4ca-4347-9052-321b30119dbe">

Let me hand this off to @taras-yemets to take care of, as part of migrating the "Manage Subscription" UI to Sourcegraph.com.

Essentially the work to do here boils down to this:
- Look at the wonderful design Rob put together [here](https://github.com/sourcegraph/cody/issues/3539#issue-2206152836)
- Add that top banner/image as applicable
- Add that Error call-out and border as applicable

The condition under which we show this error state is IFF `subscriptionDetails.status === 'past-due'`. That is, if the current Cody Pro subscription is in the PastDue state.

We want to nudge the user to update their payment method on-file. But once they do that, they will just need to wait until we automatically retry the failed payment again. (Which may be hours or even days later, depending on how long it has been since the last payment attempt.)

If we want to update the logic to automatically try to charge the "new" credit card immediately after the user has supplied a new payment method, we should track that as a separate issue. (Since it would require updating both the frontend and backend to account for that... and I probably wouldn't be a big priority to do anytime soon.)
